### ISSUE TITLE
native_batch_norm_backward extremely slow on MPS for 3D tensors (slower than CPU)

### ISSUE THREAD

**Issue by cefalix** [pytorch/pytorch#178492]
## 🐛 Describe the bug

  3D CNN training on MPS is **slower than CPU** due to `native_batch_norm_backward` consuming ~70% of training time for 5D tensors (batch, channels, D, H, W).
  2D training on MPS works excellently (9–12× speedup), so this appears specific to the 3D batch norm backward kernel.

  ### Profiling result

  The bottleneck is `aten::native_batch_norm_backward` for 3D (5D) tensors on MPS. Forward pass and all other 3D ops run at reasonable speed.

  ### Benchmark results

  | Test | Device | Time (s) | Throughput | Speedup |
  |------|--------|-----------|------------|---------|
  | 2D ResNet18 | CPU | 45.7 | 21 img/s | — |
  | 2D ResNet18 | MPS | 5.1 | 188 img/s | **8.9×** |
  | 3D ResNet18 | CPU | 119.8 | 1.0 vol/s | — |
  | 3D ResNet18 | MPS | 154.9 | 0.8 vol/s | **0.77×** (slower!) |

  Tested on both PyTorch 2.7.1 and 2.10.0 — no improvement in 3D MPS performance.

  ## Reproducer

  ```python
  import time
  import torch
  import torch.nn as nn

  assert torch.backends.mps.is_available(), "MPS not available"

  # Minimal 3D model with BatchNorm3d
  class Simple3DCNN(nn.Module):
      def __init__(self):
          super().__init__()
          self.net = nn.Sequential(
              nn.Conv3d(1, 32, 3, padding=1),
              nn.BatchNorm3d(32),
              nn.ReLU(),
              nn.Conv3d(32, 64, 3, padding=1),
              nn.BatchNorm3d(64),
              nn.ReLU(),
              nn.AdaptiveAvgPool3d(1),
              nn.Flatten(),
              nn.Linear(64, 4),
          )
      def forward(self, x):
          return self.net(x)

  def bench(device_name, steps=20):
      device = torch.device(device_name)
      model = Simple3DCNN().to(device)
      optimizer = torch.optim.Adam(model.parameters())
      criterion = nn.CrossEntropyLoss()
      x = torch.randn(4, 1, 64, 64, 64, device=device)
      y = torch.randint(0, 4, (4,), device=device)

      # Warmup
      for _ in range(3):
          loss = criterion(model(x), y)
          loss.backward()
          optimizer.step()
          optimizer.zero_grad()
      if device_name == "mps":
          torch.mps.synchronize()

      start = time.perf_counter()
      for _ in range(steps):
          loss = criterion(model(x), y)
          loss.backward()
          optimizer.step()
          optimizer.zero_grad()
      if device_name == "mps":
          torch.mps.synchronize()
      elapsed = time.perf_counter() - start
      print(f"{device_name.upper():>4s}: {elapsed:.2f}s ({4*steps/elapsed:.1f} vol/s)")

  bench("cpu")
  bench("mps")

  Expected behavior

  MPS should be faster than (or at least comparable to) CPU for 3D batch norm backward, similar to how 2D batch norm achieves excellent MPS speedups.

  Environment

  - PyTorch: 2.10.0 (also tested 2.7.1 — same behavior)
  - OS: macOS 26.1 / 26.2
  - Hardware: MacBook Pro M1 Pro (32 GB) and Mac mini M4 Pro (64 GB) — same issue on both
  - Python: 3.12

  cc: #154052 (context from the "most requested MPS ops" discussion)

  ---## 🐛 Describe the bug

  3D CNN training on MPS is **slower than CPU** due to `native_batch_norm_backward` consuming ~70% of training time for 5D tensors (batch, channels, D, H, W).
  2D training on MPS works excellently (9–12× speedup), so this appears specific to the 3D batch norm backward kernel.

  ### Profiling result

  The bottleneck is `aten::native_batch_norm_backward` for 3D (5D) tensors on MPS. Forward pass and all other 3D ops run at reasonable speed.

  ### Benchmark results

  | Test | Device | Time (s) | Throughput | Speedup |
  |------|--------|-----------|------------|---------|
  | 2D ResNet18 | CPU | 45.7 | 21 img/s | — |
  | 2D ResNet18 | MPS | 5.1 | 188 img/s | **8.9×** |
  | 3D ResNet18 | CPU | 119.8 | 1.0 vol/s | — |
  | 3D ResNet18 | MPS | 154.9 | 0.8 vol/s | **0.77×** (slower!) |

  Tested on both PyTorch 2.7.1 and 2.10.0 — no improvement in 3D MPS performance.

  ## Reproducer

  ```python
  import time
  import torch
  import torch.nn as nn

  assert torch.backends.mps.is_available(), "MPS not available"

  # Minimal 3D model with BatchNorm3d
  class Simple3DCNN(nn.Module):
      def __init__(self):
          super().__init__()
          self.net = nn.Sequential(
              nn.Conv3d(1, 32, 3, padding=1),
              nn.BatchNorm3d(32),
              nn.ReLU(),
              nn.Conv3d(32, 64, 3, padding=1),
              nn.BatchNorm3d(64),
              nn.ReLU(),
              nn.AdaptiveAvgPool3d(1),
              nn.Flatten(),
              nn.Linear(64, 4),
          )
      def forward(self, x):
          return self.net(x)

  def bench(device_name, steps=20):
      device = torch.device(device_name)
      model = Simple3DCNN().to(device)
      optimizer = torch.optim.Adam(model.parameters())
      criterion = nn.CrossEntropyLoss()
      x = torch.randn(4, 1, 64, 64, 64, device=device)
      y = torch.randint(0, 4, (4,), device=device)

      # Warmup
      for _ in range(3):
          loss = criterion(model(x), y)
          loss.backward()
          optimizer.step()
          optimizer.zero_grad()
      if device_name == "mps":
          torch.mps.synchronize()

      start = time.perf_counter()
      for _ in range(steps):
          loss = criterion(model(x), y)
          loss.backward()
          optimizer.step()
          optimizer.zero_grad()
      if device_name == "mps":
          torch.mps.synchronize()
      elapsed = time.perf_counter() - start
      print(f"{device_name.upper():>4s}: {elapsed:.2f}s ({4*steps/elapsed:.1f} vol/s)")

  bench("cpu")
  bench("mps")

  Expected behavior

  MPS should be faster than (or at least comparable to) CPU for 3D batch norm backward, similar to how 2D batch norm achieves excellent MPS speedups.

  Environment

  - PyTorch: 2.10.0 (also tested 2.7.1 — same behavior)
  - OS: macOS 26.1 / 26.2
  - Hardware: MacBook Pro M1 Pro (32 GB) and Mac mini M4 Pro (64 GB) — same issue on both
  - Python: 3.12

  cc: #154052 (context from the "most requested MPS ops" discussion)

  ---

### Versions

native_batch_norm_backward extremely slow on MPS for 3D tensors (slower than CPU)

cc @jerryzh168 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by benediktjohannes** [comment]
I guess that the „reproducer“ should not include texts down below and the „reproducer“ itself again, should it?

**Comment by hvaara** [comment]
Looks like it might have been copy-pasted a few times. `🐛 Describe the bug` is mentioned 3 times in the description. @cefalix
 do you mind cleaning up the description?

I'm not able to reproduce on M3 Max 128GB with PyTorch 2.10.0. Seeing about a 3x speedup on MPS:

```
 CPU: 20.18s (4.0 vol/s)
 MPS: 6.42s (12.5 vol/s)
```

**Comment by benediktjohannes** [comment]
@hvaara thank you for your feedback! One question for further Investigation: How much speedup in comparison between CPU and MPS do you See for 2D ResNet18? Because if this is maybe 30x more then the other Problem may still persist. Id hope that this also brings about x8 speedup, but lets Investigate on this, looking forward to seeing the results! Thanks!

**Comment by qflen** [comment]
The headline `native_batch_norm_backward` issue was fixed by #180335 (commit 50294ed45005, 2026-04-22), which flattens 5D batch_norm to 4D before dispatch. That PR closed #180334 but didn't reference this one.

OP's repro on current `main` (b16f9063b2), Apple Silicon:

```
3D, x=[4,1,64,64,64], 20 steps
  CPU 19.90s   MPS 11.34s   (MPS 1.75x faster; OP saw 0.77x slower)

2D, x=[4,1,512,512], 20 steps (same numel)
  CPU  6.61s   MPS  2.59s   (MPS 2.55x faster)
```

3D-vs-2D ratios at same numel:

|  | M5 post #180335 | M3 Max pre-fix | CUDA T4 |
|---|---|---|---|
| MPS 3D / MPS 2D | 4.38x | 9.4x | 2.76x |
| CPU 3D / CPU 2D | 3.01x | n/a | n/a |

Half of @hvaara's MPS-vs-CUDA gap is gone. The residual is mostly Conv3d (27 multiplies per output vs Conv2d's 9, so ~3x is fundamental):

| Op (fwd+bwd, 3D step) | Time | % step |
|---|---|---|
| Conv3d 32→64 | 378 ms | ~66% |
| Conv3d 1→32 | 132 ms | ~23% |
| BN3d C=64 | 26 ms | ~5% |
| AdaptiveAvgPool3d→1 | 26 ms | ~5% |
| ReLU (5D) | 18 ms | ~3% |
| BN3d C=32 | 13 ms | ~2% |

`native_batch_norm_backward` is no longer hot. Filed #182823 for the residual Conv3d gap.

@cefalix could you retest on M1 Pro / M4 Pro on `main`? 

I think this issue can be closed.
